satellite/accounting/live/live_test.go

```diff
@@ -14,35 +14,188 @@ import (
+	}
+}
+
+func TestRedisLiveAccounting(t *testing.T) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	address, cleanup, err := redisserver.Start()
+	require.NoError(t, err)
+	defer cleanup()
+
+	config := Config{
+		StorageBackend: "redis://" + address + "?db=0",
+	}
+	cache, err := NewCache(zaptest.NewLogger(t).Named("live-accounting"), config)
+	require.NoError(t, err)
+
+	// ensure we are using the expected underlying type
+	_, ok := cache.(*redisLiveAccounting)
+	assert.True(t, ok)
+
+	projectIDs, sum, err := populateCache(ctx, cache)
+	require.NoError(t, err)
+
+	// make sure all of the "projects" got all space updates and got right totals
+	for _, projID := range projectIDs {
+		spaceUsed, err := cache.GetProjectStorageUsage(ctx, projID)
+		require.NoError(t, err)
+		assert.Equalf(t, sum, spaceUsed, "projectID %v", projID)
+	}
+
+	err = cache.ResetTotals(ctx)
+	require.NoError(t, err)
+
+	for _, projID := range projectIDs {
+		spaceUsed, err := cache.GetProjectStorageUsage(ctx, projID)
+		require.NoError(t, err)
+		assert.EqualValues(t, 0, spaceUsed)
+	}
+}
+
+func TestRedisCacheConcurrency(t *testing.T) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	address, cleanup, err := redisserver.Start()
+	require.NoError(t, err)
+	defer cleanup()
+
+	config := Config{
+		StorageBackend: "redis://" + address + "?db=0",
+	}
+	cache, err := NewCache(zaptest.NewLogger(t).Named("live-accounting"), config)
+	require.NoError(t, err)
+
+	projectID := testrand.UUID()
+
 	const (
-		valuesListSize  = 1000
-		valueMultiplier = 4096
-		numProjects     = 200
+		numConcurrent = 100
+		inlineAmount  = 10
+		remoteAmount  = 10
 	)
+	expectedSum := (inlineAmount * numConcurrent) + (remoteAmount * numConcurrent)
+
+	var group errgroup.Group
+	for i := 0; i < numConcurrent; i++ {
+		group.Go(func() error {
+			return cache.AddProjectStorageUsage(ctx, projectID, inlineAmount, remoteAmount)
+		})
+	}
+	require.NoError(t, group.Wait())
+
+	spaceUsed, err := cache.GetProjectStorageUsage(ctx, projectID)
+	require.NoError(t, err)
+
+	require.EqualValues(t, expectedSum, spaceUsed)
+}
+
+func TestPlainMemoryCacheConcurrency(t *testing.T) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	config := Config{
+		StorageBackend: "plainmemory",
+	}
+	cache, err := NewCache(zaptest.NewLogger(t).Named("live-accounting"), config)
+	require.NoError(t, err)
+
+	projectID := testrand.UUID()
+
+	const (
+		numConcurrent = 100
+		inlineAmount  = 10
+		remoteAmount  = 10
+	)
+	expectedSum := (inlineAmount * numConcurrent) + (remoteAmount * numConcurrent)
+
+	var group errgroup.Group
+	for i := 0; i < numConcurrent; i++ {
+		group.Go(func() error {
+			return cache.AddProjectStorageUsage(ctx, projectID, inlineAmount, remoteAmount)
+		})
+	}
+	require.NoError(t, group.Wait())
+
+	spaceUsed, err := cache.GetProjectStorageUsage(ctx, projectID)
+	require.NoError(t, err)
+
+	require.EqualValues(t, expectedSum, spaceUsed)
+}
+
+func TestResetTotals(t *testing.T) {
```

Comment:
I'm wondering if this name is appropriated for this test. It's called `ResetTotals` but I don't see the call to it.

Reply:
Yes, I'm not sure what this test is for

Reply:
It looks that this function is already tested in the `Redis` and `Plainmemory` tests, so we should remove this one, unless that we want a test which passes a negative number to `AddProjectStorageUsage` (https://github.com/storj/storj/pull/3213/files/6ba21dabef6135a3645541d4d7dfa8e76ddf22ed#diff-72f4bb03961ddf4ea47393d293e27c2fR180), in that case, we should rename it.

Reply:
Hmm, looking at the test now, I see that it is actually possible to subtract from the space used by passing a negative number into `AddProjectStorageUsage`, which I don't think we want to do. I'll add some checks and use this test to ensure that we can't pass negative numbers into the function